feat: add email filtering to team memberships endpoint

Devanshusharma2005 opened the pull request:
Fixes ##22658

## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Fixes #XXXX (GitHub issue number)
- Fixes CAL-XXXX (Linear issue number - should be visible at the bottom of the GitHub issue description)

## Visual Demo (For contributors especially)

A visual demonstration is strongly recommended, for both the original and new change **(video / image - any one)**.

#### Video Demo (if applicable):

- Show screen recordings of the issue or feature.
- Demonstrate how to reproduce the issue, the behavior before and after the change.

#### Image Demo (if applicable):

- Add side-by-side screenshots of the original and updated change.
- Highlight any significant change(s).

## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Checklist

<!-- Remove bullet points below that don't apply to you -->

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

Devanshusharma2005 committed:
Controller Layer updates.

Devanshusharma2005 committed:
adding email filtering and pagination to team memberships endpoint..

Devanshusharma2005 commented on apps/api/v2/src/modules/teams/memberships/controllers/teams-memberships.controller.ts:
```diff
@@ -85,16 +85,20 @@ export class TeamsMembershipsController {
+    @Query() queryParams: GetUsersInput
   ): Promise<GetTeamMembershipsOutput> {
-    const { skip, take } = queryParams;
+    const { skip, take, emails } = queryParams;
```
Now extracting emails also for the filtering.

Devanshusharma2005 committed:
Minor enhancements.

Devanshusharma2005 committed:
Merge branch 'main' into feat/teams-memberships-filter-by-email

supalarry reviewed (commented):
Good job! Some comments below:

Devanshusharma2005 committed:
Improve addressed.

Devanshusharma2005 committed:
Merge branch 'feat/teams-memberships-filter-by-email' of https://github.com/Devanshusharma2005/cal into feat/teams-memberships-filter-by-email

Devanshusharma2005 committed:
Merge branch 'main' into feat/teams-memberships-filter-by-email

Devanshusharma2005 committed:
Merge branch 'main' into feat/teams-memberships-filter-by-email

supalarry commented on apps/api/v2/src/modules/teams/memberships/inputs/get-team-memberships.input.ts:
```diff
@@ -0,0 +1,23 @@
+import { Transform } from "class-transformer";
+import { IsOptional, Validate, ArrayMaxSize } from "class-validator";
+
+import { GetUsersInput } from "../../../users/inputs/get-users.input";
```
It's better to use absolute path `"@/modules/users/inputs/get-users.input";` and same for `IsEmailStringOrArray` unless the file is in the same folder.

supalarry commented on apps/api/v2/src/modules/teams/memberships/controllers/teams-memberships.controller.e2e-spec.ts:
```diff
@@ -328,6 +328,131 @@ describe("Teams Memberships Endpoints", () => {
       return request(app.getHttpServer()).get(`/v2/teams/${team.id}/memberships/123132145`).expect(404);
     });
 
+    it("should filter memberships by single email", async () => {
```
Nice tests!

supalarry commented on apps/api/v2/src/modules/teams/memberships/inputs/get-team-memberships.input.ts:
```diff
@@ -0,0 +1,23 @@
+  @Transform(({ value }: { value: string | string[] }) => {
+    return typeof value === "string" ? [value] : value;
+  })
+  @ApiPropertyOptional({
```
I think that e2e test `it("should filter memberships by multiple emails"` will fail because there pass emails as comma separated string but here we expect it as array in url. I see that logic comes from `GetOrganizationsUsersInput` - i did not see that someone added it, but in other places we have it as comma separated string that then we turn into array, so you need to update `emails` to be like we have in `apps/api/v2/src/modules/oauth-clients/controllers/oauth-client-users/inputs/get-managed-users.input.ts` aka we need something like
```
@IsOptional()
  @Transform(({ value }) => {
    if (typeof value === "string") {
      return value.split(",").map((email: string) => {
        const trimmed = email.trim();
        if (!isEmail(trimmed)) {
          throw new BadRequestException(`Invalid email ${trimmed}`);
        }
        return trimmed;
      });
    }
    return value;
  })
  @ArrayNotEmpty({ message: "emails cannot be empty." })
  @ApiPropertyOptional({
    description:
      "Filter managed users by email. If you want to filter by multiple emails, separate them with a comma.",
    example: "?emails=email1@example.com,email2@example.com",
  })
  emails?: string[];
```

here

supalarry reviewed (commented):
Good job we are almost there : )

Devanshusharma2005 committed:
refactor: update team memberships input to handle comma-separated emails

- Replace array format with comma-separated string handling
- Add proper email validation with BadRequestException
- Remove ArrayMaxSize constraint for better flexibility
- Update API documentation and examples
- Align with codebase patterns from get-managed-users.input.ts

Devanshusharma2005 committed:
Merge branch 'main' into feat/teams-memberships-filter-by-email

supalarry reviewed (dismissed):
Great work! 🚀

Devanshusharma2005 committed:
Merge branch 'main' into feat/teams-memberships-filter-by-email

ThyMinimalDev commented on apps/api/v2/src/modules/teams/memberships/inputs/get-team-memberships.input.ts:
```diff
@@ -0,0 +1,29 @@
+    if (typeof value === "string") {
+      return value.split(",").map((email: string) => {
+        const trimmed = email.trim();
+        if (!isEmail(trimmed)) {
+          throw new BadRequestException(`Invalid email ${trimmed}`);
+        }
```
not a fan of mixing validation logic with transform, use @IsEmail decorator instead

ThyMinimalDev replied:
```
@IsEmail({},{each: true})
```

Devanshusharma2005 committed:
Morgan suggestions addressed.

Devanshusharma2005 committed:
Merge branch 'main' into feat/teams-memberships-filter-by-email

ThyMinimalDev commented on apps/api/v2/src/modules/teams/memberships/inputs/get-team-memberships.input.ts:
```diff
@@ -0,0 +1,23 @@
+import { ArrayNotEmpty, IsEmail, IsOptional } from "class-validator";
+
+export class GetTeamMembershipsInput extends GetUsersInput {
+  @IsOptional()
+  @Transform(({ value }) => {
+    if (typeof value === "string") {
+      return value.split(",").map((email: string) => email.trim());
+    }
+    return value;
+  })
+  @ArrayNotEmpty({ message: "emails cannot be empty." })
+  @IsEmail({}, { each: true, message: "Each email must be a valid email address" })
+  @ApiPropertyOptional({
+    type: [String],
+    description:
+      "Filter team memberships by email addresses. If you want to filter by multiple emails, separate them with a comma (max 20 emails for performance).",
+    example: "?emails=user1@example.com,user2@example.com",
+  })
+  emails?: string[];
```
and use ArrayMaxSize to enforce the 20 emails

Devanshusharma2005 committed:
More improvements......

Devanshusharma2005 committed:
Merge branch 'main' into feat/teams-memberships-filter-by-email

Devanshusharma2005 committed:
Merge branch 'main' into feat/teams-memberships-filter-by-email
